repentogon/Patches/ASMPatches/ASMStatusEffects.cpp

```diff
@@ -0,0 +1,169 @@
+*/ /////////////////////
+
+void ASMPatchInlinedGetStatusEffectTarget(void* addr, ASMPatch::Registers entityReg, ASMPatch::Registers targetReg, ASMPatch::SavedRegisters::Registers savedReg, unsigned int jumpOffset) {
+	ASMPatch::SavedRegisters savedRegisters(ASMPatch::SavedRegisters::Registers::GP_REGISTERS + ASMPatch::SavedRegisters::XMM1 - savedReg, true);
+	ASMPatch patch;
+	patch.PreserveRegisters(savedRegisters)
+		.Push(entityReg)
+		.AddInternalCall(GetStatusEffectTargetTrampoline)
+		// poor man's mov reg,reg
+		.Push(ASMPatch::Registers::EAX)
+		.Pop(targetReg)
+		.RestoreRegisters(savedRegisters)
+		.AddRelativeJump((char*)addr + jumpOffset);
+	sASMPatcher.PatchAt(addr, &patch);
+}
+
+/* /////////////////////
+// Thank you, God bless you, and God bless the United States of America.
+*/ /////////////////////
+
+bool avoidIceCrash = false;
+void __stdcall AvoidIceCrashHack() {
+	avoidIceCrash = true;
+}
+
+void ASMPatchIcedKnockbackPreNPC() {
+	SigScan scanner("6a0168c80000008d85");
+	scanner.Scan();
+	void* addr = scanner.GetAddress();
+
+	ASMPatch patch;
+	patch.AddInternalCall(AvoidIceCrashHack)
+		.AddBytes("\x6a\x01\x68\xc8").AddZeroes(3)
+		.AddRelativeJump((char*)addr + 7);
+	sASMPatcher.PatchAt(addr, &patch);
+}
+
+void ASMPatchIcedKnockbackPrePlayer() {
+	SigScan scanner("68c8000000508d4424");
+	scanner.Scan();
+	void* addr = scanner.GetAddress();
+
+	ASMPatch patch;
+	patch.AddInternalCall(AvoidIceCrashHack)
+		.AddBytes("\x68\xc8").AddZeroes(3)
+		.AddRelativeJump((char*)addr + 5);
+	sASMPatcher.PatchAt(addr, &patch);
+}
+
+HOOK_METHOD(Entity, AddKnockback, (EntityRef& ref, const Vector& pushDirection, int duration, bool takeImpactDamage) -> void) {
+	if (!avoidIceCrash)
+		super(ref, pushDirection, duration, takeImpactDamage);
+	else
+	{
+		Entity* entity = this;
+		if (!IgnoreEffectFromFriendly(&ref)) {
+			if (GetStatusEffectTarget()) {
+				Entity* parent = this->_parent;
+				Entity* otherEnt = parent;
+				while (otherEnt = parent, otherEnt != nullptr) {
+					entity = otherEnt;
+					parent = otherEnt->_parent;
+				}
+			}
+			if (duration != 0 && ((entity->_flags & 1) == 0 || (entity->_flags & 0x2000000000000) != 0)) {
+				entity->_knockbackCountdown = duration;
+				entity->_knockbackDirection = pushDirection;
+				if (10.0f < entity->_mass && ((this->_flags & 0x2000000000000) == 0)) {
+					float logMass = 1.0f / std::log10(entity->_mass);
+					entity->_knockbackDirection *= logMass;
+					int countdown = (int)(entity->_knockbackCountdown * logMass);
+					if (countdown < 1)
+						countdown = 1;
+					entity->_knockbackCountdown = countdown;
+				}
+				entity->_flags |= 0x800000000000;
+				if (takeImpactDamage)
+					entity->_flags |= 0x1800000000000;
+				entity->CopyStatusEffects();
+			}
+		}
+
+		avoidIceCrash = false;
+	}
+}
+
+/* /////////////////////
+// Run patches
+*/ /////////////////////
+
+void PatchInlinedGetStatusEffectTarget()
+{
+	ZHL::Logger logger;
+	for (StatusEffectPatchInfo i : patches) {
```

Comment:
Add const& to capture by reference, will improve perf and avoid unexpected stuff like weird copies